pylint/config/option_manager_mixin.py

```diff
@@ -311,6 +294,35 @@ def read_config_file(self, config_file=None, verbose=None):
             msg = "No config file found, using default configuration"
         print(msg, file=sys.stderr)
 
+    @staticmethod
+    def _parse_toml(
+        config_file: Union[Path, str], parser: configparser.ConfigParser
```

Comment:
I'm trying to use the largest possible type here because ``open`` can take both ``str`` or ``Path`` so using this typing is true and permit to not cast to string for no reason when we use a Path before calling.

Reply:
It is just from experience that I know that allowing both `Path` and `str` can create some problems later on. 
I think it was the typing of `file_location` or something in which initially both `Path` and `str` were allowed, which required a lot of `isinstance()`'s only to later find out that it couldn't ever be `Path`. I think in this case it should be fine, but I have noticed that we spare ourselves some headaches if we normalise all `Path`'s to be `str` when we are handling and passing paths around internally.

Reply:
> required a lot of isinstance()'s

I guess it's to stop mypy from complaining ? Path(str) or str(path) should work to go from one to the other (if the path can't be None) so I would cast directly instead of using isinstance ? But I agree that if we have some function that require one or the other casting to string/path is indeed a lot of work.

In the current function ``Union[Path, str]`` is not a "lie".  But If we want to be able to use Path some day I think using the real typing we can handle when we can is important. It prevents having to cast to str to satisfy mypy.

Reply:
I wonder if there is every any reason to use `Path`. It seems the only difference between `Path` and `str` is that `Path` can be a `bytes` and can be easily converted between `str` and `bytes` in both ways.
See:
https://docs.python.org/3/glossary.html#term-path-like-object

I might be wrong here, but I don't immediately see why we would ever need it to be `Path`. 
Besides, I think being able to know that pathnames within `pylint` are always `str` makes it a lot easier to handle them and write code that accepts it. That level of consistency makes the codebase more maintainable imo.

Reply:
There's 17 files using ``pathlib.Path`` already (I'm probably mostly the one who added them). It's personal preference but I rather do ``my_path.resolve() / "directory" / "file.py"`` than ``os.path.abspath(os.path.join(my_path, "directory", "file.py))`` or ``Path(__file__).resolve().parent.parent`` than ``os.path.dirname(os.path.dirname(os.path.abspath(__file__)))``. Also ``mypath.glob('*.csv')`` does not require an additional import to work. Last thing, Path is more semantic, a string can be something else than a path so using Path indicate intent.

I'm not on a crusade to change everything to Path everywhere, but I think being able to use Path would be convenient. Also all python functions handle both, so if we can't, it's probably an internal pylint problem. Putting the real typing in new function is the way to slowly fix this problem, so I think we should do it. Putting only str will raise mypy error that are not really errors (which might be the reason you had so much trouble with Path vs str in the first place ?).

Reply:
Yeah you convinced me, we should allow both here.

I looked at what the initial problem was in https://github.com/PyCQA/pylint/pull/4980/files. Seems like this was because of a similar discussion: should we use `Union[str, Path]` or should we type like we are currently using the functions (`str`)?. In that case we decided against allowing `Path` because it would mean a lot of additional typing and some refactoring. Here it might be fine to do allow it as we are also not returning the `parameter` that is typed as `Path`.